enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,284 @@
+By specifying a `network-hint`, only a node that has an interface in that CIDR can be selected for the service's traffic (as explained earlier), and the service's pods egress traffic will exit through that interface.
+By not specifying a `network-hint` any node in the cluster can be chosen to manage the service's traffic, and the traffic will exit according to the node's routing table.
+
+### Implementation Details/Notes/Constraints
+
+A new annotation `k8s.ovn.org/egress-service` is supported for LoadBalancer services.
+When `ovnkube-master` detects that a LoadBalancer service has this annotation it will elect a node to act as the point for all of the traffic of that service (ingress/egress). If the annotation contains a valid CIDR in its `network-hint` only a node that has an interface in that CIDR can be elected, which is possible by inspecting all of the nodes' `k8s.ovn.org/host-addresses` annotation.
+If the `network-hint` is not specified any node can be elected.
+After choosing a node, it will create a logical router policy on the ovn_cluster_router for all of the endpoints of the service to steer their egress traffic to that node's mgmt port.
+We should take extra care with these policies to not break pod to pod traffic and incoming service traffic, thus we should not steer the traffic when the destination is a pod (pod to pod) or an rtoj port (reply to service traffic).
+
+For example when 10.244.0.3 and 10.244.1.6 are the endpoints of the service and the elected node's mgmt port is 10.244.0.2 we expect policies like these to be created:
+```none
+$ ovn-nbctl lr-policy-list ovn_cluster_router
+       ...
+       200           ip4.src == 10.244.0.3/32 && ip4.dst != 100.64.0.0/16 && ip4.dst != 10.244.0.0/16       reroute      10.244.0.2
+       200           ip4.src == 10.244.1.6/32 && ip4.dst != 100.64.0.0/16 && ip4.dst != 10.244.0.0/16       reroute      10.244.0.2
+       ...
+```
+After that the service will be annotated with `k8s.ovn.org/egress-service-host=<node_name>` and the node labeled with `k8s.ovn.org/<svc-name>-<svc namespace>-host: ""`.
+
+When `ovnkube-node` detects that a LoadBalancer service has the `k8s.ovn.org/egress-service` annotation and it is running in the node specified in the service's `k8s.ovn.org/egress-service-host` annotation, it will add the relevant SNATs to the host's iptables. If the `network-hint` is specified, it will find the interface associated with the corresponding CIDR and use it to exit the host.
+
+For example when the `eth1` interface on ovn-worker belongs to the `network-hint` CIDR, 10.244.0.3 and 10.244.1.6 are the endpoints of the annotated LoadBalancer service whose external IP is 172.19.0.100, we expect to see iptables rules like these in the "nat" table of `ovn-worker`:
+```none
+$ iptables-save
+...
+-A POSTROUTING -j OVN-KUBE-EGRESS-SVC
+...
+-A OVN-KUBE-EGRESS-SVC -s 10.244.1.6/32 -o eth1 -j SNAT --to-source 172.19.0.100
```

Comment:
sorry, not sure I understand what effect this has here? the man page states:
```
Name of an interface via which a packet is going to be sent (for packets entering the FORWARD, OUTPUT and POSTROUTING chains)..
```

Reply:
IPTables rules can't directly affect the packet routing. The docs are confusing because they suffer from a combination of being written by people who didn't actually understand the topic, and being written by people who understand the topic so well that they can't describe it in terms that make sense to someone who doesn't already understand it. :slightly_smiling_face:. But "sets the outgoing network interface for a rule" is meant to be interpreted as "sets the 'outgoing interface' value that this rule will match".

If you need to force some packets out eth1 despite eth0 having an apparently-better route for them, then you'd need to use policy-based routing. eg, https://superuser.com/questions/376667/how-to-route-only-specific-subnet-source-ip-to-a-particular-interface. That would be based on the original source IP, not the SNAT IP, since the SNAT is being done from `POSTROUTING`, which (spoiler alert!) happens after routing.

Reply:
changed to using fwmarks and custom routing tables, does it make sense now?